### PR TITLE
The string schism

### LINKED ISSUE #12: The EDSL needs a disctinction between Strings and C-Strings
I call C-Strings the ones that cannot contain `\x00`.

The problem arises too often.

### PR DESCRIPTION
**Pull request by smondet**
- finally do the `c_string` Vs `byte_array` schism (fixes #12)
- switch to `jbuilder`
- revamp the custom testing framework

### CODE REVIEW

File: genspio.opam [hunk 1]
```diff
@@ -1,31 +1,22 @@
...
+version: "0.0.1-dev"
+available: [
+  ocaml-version >= "4.03.0"
 ]
-
-install: [
-  [make "META"]
-  [make "genspio.install"]
+build: [
+  ["ocaml" "please.ml" "configure"]
```
Comment (rleonid) [inline genspio.opam]: oh the horror...

File: src/lib/EDSL.mli [hunk 2]
```diff
@@ -4,15 +4,19 @@
 type 'a t = 'a Language.t
 (** The type of a Genspio expression. *)
 
+type c_string
+type byte_array
+
```
Comment (rleonid) [inline src/lib/EDSL.mli]: Please add documentation

File: src/lib/EDSL.mli [hunk 3]
```diff
@@ -106,26 +116,29 @@ val list_append: 'a list t -> 'a list t -> 'a list t
...
 (** [str >> cmd] is [feed ~string:str cmd]. *)
 
-val string_concat: string t list -> string t
-(** Concatenate an (OCaml) list of [string t] values. *)
+val string_concat: c_string t list -> c_string t
+(** Concatenate an (OCaml) list of [c_string t] values. *)
 
-val string_concat_list: string list t -> string t
-(** Concatenate a Genspio list of strings [string list t]. *)
+val string_concat_list: c_string list t -> c_string t
```
Comment (rleonid) [inline src/lib/EDSL.mli]: Maybe it is time to wrap some of these functions into a List module like with Integers above?

File: genspio.opam (hunk 1 above)
Comment (smondet) [reply genspio.opam]: Quite proud of myself there :)

File: src/lib/EDSL.mli (hunk 2 above)
Comment (smondet) [reply src/lib/EDSL.mli]: OK :+1:

File: src/lib/EDSL.mli (hunk 3 above)
Comment (smondet) [reply src/lib/EDSL.mli]: Yes I plan to revamp the API quite a bit but not in this PR.

**Comment by smondet** [comment]
@leo the necessity of the split is in the issue #12 and the talk and it all comes down to this piece of C code: http://wr.mondet.org/slides/OCaml2017-Genspio/#/c-strings-vs-byte-arrays

**Pull request merged**